include/alpaka/rand/Philox/PhiloxBaseCommon.hpp

```diff
@@ -0,0 +1,90 @@
+             * @tparam TWidth width of one counter element (in bits)
+             * @tparam TRounds number of S-box rounds
+             */
+            // TODO: Not sure where to put this
+            template<unsigned TCounterSize, unsigned TWidth, unsigned TRounds>
+            struct PhiloxParams
+            {
+                static unsigned constexpr counterSize = TCounterSize;
+                static unsigned constexpr width = TWidth;
+                static unsigned constexpr rounds = TRounds;
+            };
+
+            /** Common class for Philox family engines
+             *
+             * Checks the validity of passed-in parameters and calls the \a TBackend methods to perform N rounds of the
+             * Philox shuffle.
+             *
+             * @tparam TBackend device-dependent backend, specifies the array types
+             * @tparam TParams Philox algorithm parameters \sa PhiloxParams
+             * @tparam TImpl engine type implementation (CRTP)
+             */
+            template<typename TBackend, typename TParams, typename TImpl>
+            class PhiloxBaseCommon : public TBackend
+            {
+                static unsigned const numRounds = TParams::rounds;
+                static unsigned const vectorSize = TParams::counterSize;
+                static unsigned const numberWidth = TParams::width;
+
+                static_assert(numRounds > 0, "Number of Philox rounds must be > 0.");
+                static_assert(vectorSize % 2 == 0, "Philox counter size must be an even number.");
+                static_assert(vectorSize <= 16, "Philox SP network is not specified for sizes > 16.");
+                static_assert(numberWidth % 8 == 0, "Philox number width in bits must be a multiple of 8.");
+
+                // static_assert(TWidth == 32 || TWidth == 64, "Philox implemented only for 32 and 64 bit numbers.");
+                static_assert(numberWidth == 32, "Philox implemented only for 32 bit numbers.");
+
+            public:
+                using Counter = typename TBackend::Counter;
+                using Key = typename TBackend::Key;
+
+            protected:
+                /** Performs N rounds of the Philox shuffle
+                 *
+                 * @param counter_in initial state of the counter
+                 * @param key_in initial state of the key
+                 * @return result of the PRNG shuffle; has the same size as the counter
+                 */
+                ALPAKA_FN_HOST_ACC auto nRounds(Counter const& counter_in, Key const& key_in) -> Counter
+                {
+                    Key key{key_in};
+                    Counter counter = TBackend::singleRound(counter_in, key);
+                    helpers::repeat_n<numRounds>()(
+                        [&]()
+                        {
+                            key = TBackend::bumpKey(key);
+                            counter = TBackend::singleRound(counter, key);
+                        });
```

Comment:
Why can't this be a `for` loop? Potentially with `#pragma unroll`. This looks like a premature optimization.

Reply:
I did it becuse all the other Philox implementations do it as well. I'll run some performance tests when the PR is ready.